cidc_api/services/files.py

```diff
@@ -0,0 +1,36 @@
+"""Endpoints for downloadable file operations"""
+
+from multiprocessing.pool import ThreadPool
+
+from eve import Eve
+
+import gcloud_client
+
+
+def register_files_hooks(app: Eve):
+    app.on_fetched_resource_downloadable_files += insert_download_urls
+    app.on_fetched_item_downloadable_files += insert_download_url
+
+
+def insert_download_url(payload: dict):
+    """
+    Get a signed GCS download URL for the requested file.
+
+    TODO: evaluate ways of caching signed URLs for files that haven't yet
+    expired, so that we aren't re-generating them on every request.
+    """
+    object_url = payload["object_url"]
+    download_url = gcloud_client.get_signed_url(object_url)
+    payload["download_link"] = download_url
+
+
+def insert_download_urls(payload: dict):
+    """
+    Get a signed GCS download URL for each of the requested files.
+    """
+    # Each call to insert_download_url generates a request
+    # to the GCS API, so we get a speed-up from multithreading:
+    # although Python dispatches the requests synchronously,
+    # the OS handles the network requests in parallel.
+    with ThreadPool() as pool:
```

Comment:
Why is this ThreadPool needed?

Reply:
Because every invocation to `insert_download_url` makes a request to the Google Cloud Storage API. If there are a lot of files in `payload`, handling each of these requests synchronously would take a very long time. By dispatching these requests to a `ThreadPool`, the network operations will be made in parallel.